src/orca-jedi/interpolator/Interpolator.cc

```diff
@@ -177,6 +178,128 @@ template void Interpolator::executeInterpolation<float>(
     const std::vector<bool> & mask,
     std::vector<double>::iterator& iter) const;
 
+/// \brief Interpolate from model space to observation space
+/// \param vars Oops variables
+/// \param inc Increment object (input)
+/// \param mask Mask vector in observation space
+/// \param result Result (output) vector in observation space
+void Interpolator::apply(const oops::Variables& vars, const Increment& inc,
+           const std::vector<bool> & mask,
+           std::vector<double>& result) const {
+  // input is inc output is result
+  const size_t nvars = vars.size();
+
+  for (size_t j=0; j < nvars; ++j) {
+    if (!inc.variables().has(vars[j])) {
+      std::stringstream err_stream;
+      err_stream << "orcamodel::Interpolator::apply varname \" "
+                 << "\" " << vars[j]
+                 << " not found in the model increment." << std::endl;
+      err_stream << "    add the variable to the increment variables and "
+                 << "add a mapping from the geometry to that variable."
+                 << std::endl;
+      throw eckit::BadParameter(err_stream.str(), Here());
+    }
+  }
+
+  const std::vector<size_t> varSizes =
+    inc.geometry()->variableSizes(vars);
+  size_t nvals = 0;
+  for (size_t jvar=0; jvar < nvars; ++jvar) nvals += nlocs_ * varSizes[jvar];
+  result.resize(nvals);
+
+  std::size_t out_idx = 0;
+  for (size_t jvar=0; jvar < nvars; ++jvar) {
+    auto gv_varname = vars[jvar].name();
+    atlas::Field tgt_field = atlasObsFuncSpace_.createField<double>(
+        atlas::option::name(gv_varname) |
+        atlas::option::levels(varSizes[jvar]));
+    interpolator_.execute(inc.incrementFields()[gv_varname], tgt_field);
+    auto field_view = atlas::array::make_view<double, 2>(tgt_field);
+    atlas::field::MissingValue mv(inc.incrementFields()[gv_varname]);
+    bool has_mv = static_cast<bool>(mv);
+    for (std::size_t klev=0; klev < varSizes[jvar]; ++klev) {
+      for (std::size_t iloc=0; iloc < nlocs_; iloc++) {
+        if (has_mv && mv(field_view(iloc, klev))) {
+          result[out_idx] = util::missingValue<double>();
+        } else {
+          result[out_idx] = field_view(iloc, klev);
+        }
+        ++out_idx;
+      }
+    }
+  }
```

Comment:
Could we replace this loop to use the templated interpolation method above for consistency? Longer term all of this will be refactored out into OOPS, but for now it would be good to limit the repetition if we can.

```suggestion
  auto res_iter = result.begin();
  for (size_t jvar=0; jvar < nvars; ++jvar) {
      executeInterpolation<double>(vars[jvar].name(), varSizes[jvar], state, mask, res_iter);
  }
```

Reply:
Trouble is this routine uses increment rather than state. The template would have to made generic to state/increment. I'm not sure how to do this. Maybe this could go on the back burner for now?

Reply:
No worries. Happy to review your mega template when the time comes.